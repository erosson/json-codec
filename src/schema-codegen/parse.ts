import * as S from 'json-schema'

export type SchemaDefinition = { name: string, body: SchemaBody }
export type SchemaBody = SchemaSimpleType | SchemaObject | SchemaArray | SchemaRef
export type SchemaSimpleType = { type: 'string' } | { type: 'number' } | { type: 'boolean' } | { type: 'null' }
export type SchemaObject = { type: 'object', properties: SchemaProp[] }
export type SchemaProp = SchemaDefinition
export type SchemaArray = { type: 'array', items: SchemaBody }
export type SchemaRef = { type: 'ref', path: string }
export type Schema = SchemaDefinition[]

function refName(ref: SchemaRef): string | null {
    if (ref.path.startsWith('#/definitions/')) {
        return ref.path.replace('#/definitions/', '')
    }
    return null
}
export function parse(schema: S.JSONSchema7): Schema {
    return [parseDefinition(schema)]
        .concat(Object.entries(schema.definitions ?? {}).map(([name, def]) => parseDefinition(def, name)))
        .filter(d => !!d) as SchemaDefinition[]
}
function parseDefinition(def: S.JSONSchema7Definition, name?: string): SchemaDefinition | null {
    if (typeof def !== 'boolean') {
        name = name ?? def.title
        if (name) {
            const body = parseBody(def)
            if (body) {
                return { name, body }
            }
        }
    }
    return null
}

function parseBody(def: S.JSONSchema7): SchemaBody | null {
    if (def.$ref) {
        return { type: 'ref', path: def.$ref }
    }
    switch (def.type) {
        case 'string':
        case 'number':
        case 'boolean':
        case 'null':
            return { type: def.type }
        case 'integer':
            return { type: 'number' }
        case 'object':
            return parseObject(def)
        case 'array':
            return parseArray(def)
        case undefined:
        default:
            return null
    }
}

function parseArray(def: S.JSONSchema7): SchemaArray | null {
    if (def.items && typeof def.items === 'object' && !Array.isArray(def.items)) {
        const body = parseBody(def.items)
        if (body) {
            return { type: 'array', items: body }
        }
    }
    return null
}
function parseObject(def: S.JSONSchema7): SchemaObject {
    return {
        type: 'object',
        properties: Object.entries(def.properties ?? {}).map(([key, body]) => {
            return parseProp(body, key)
        })
            .filter(p => !!p) as SchemaProp[]
    }
}
function parseProp(def: S.JSONSchema7Definition, name: string): SchemaProp | null {
    return parseDefinition(def, name)
}