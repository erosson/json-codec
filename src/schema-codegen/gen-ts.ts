import * as M from 'ts-morph'
import * as P from './parse'

export function genSource(parsed: P.Schema): { errors: () => string, source: string } {
    const project = new M.Project()
    const out = project.createSourceFile('./tempfile.ts')
    gen(parsed, out)
    // const errors = project.formatDiagnosticsWithColorAndContext(project.getPreEmitDiagnostics());
    function errors() {
        return project.formatDiagnosticsWithColorAndContext(project.getPreEmitDiagnostics())
    }
    const diags = project.getPreEmitDiagnostics()
    const source = out.getText()
    out.removeText()
    return { errors, source }
}
export function gen(parsed: P.Schema, out: M.SourceFile): void {
    for (let def of parsed) {
        genDef(def, out)
    }
}
function writeBody(body: P.SchemaBody): string | null {
    switch (body.type) {
        case 'string':
        case 'number':
        case 'boolean':
        case 'null':
            return body.type
        case 'ref':
            return P.refName(body) ?? null
        case 'array':
            return `Array<${writeBody(body.items)}>`
        case 'object':
            // TODO
            return null
    }
}
function genDef(def: P.SchemaDefinition, out: M.SourceFile): void {
    switch (def.body.type) {
        case 'object':
            out.addInterface({ name: def.name, properties: def.body.properties.map(p => ({ name: p.name, type: writeBody(p.body) ?? 'unknown' })) })
            return
        default:
            out.addTypeAlias({ name: def.name, type: writeBody(def.body) ?? 'unknown' })
            return
    }
}