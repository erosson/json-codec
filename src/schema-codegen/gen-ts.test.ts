import * as P from './parse'
import * as G from './gen-ts'
import schema from '../FreeCBT.schema.json'

test('gen DistortionID', () => {
    const parsed = P.parse({ ...schema.definitions.DistortionID, title: 'DistortionID' } as any)
    const { errors, source } = G.genSource(parsed)
    expect(errors).toBeFalsy()
    expect(source).toMatchSnapshot()
})

test.only('gen Distortion', () => {
    const parsed = P.parse({ ...schema.definitions.Distortion, title: 'Distortion' } as any)
    const { errors, source } = G.genSource(parsed)
    expect(errors).toBeFalsy()
    expect(source).toMatchSnapshot()
})