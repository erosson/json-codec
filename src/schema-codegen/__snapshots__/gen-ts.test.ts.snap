// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gen Distortion 1`] = `
"interface Distortion {
    slug;
    emoji;
    labelKey;
    descriptionKey;
}
"
`;

exports[`gen DistortionID 1`] = `
"type DistortionID = string;
"
`;
